File: v1/scala/crypto-square/src/main/scala/CryptoSquare.scala
class CryptoSquare {
  private val PadString = " ";

  private def transpose(segments: List[String]): List[String] =
    segments match {
      case Nil => Nil
      case x::xs => segments.map(_.padTo(x.length, PadString))
                            .transpose
                            .map(_.mkString.replaceAll(PadString, ""))
    }

  def normalizePlaintext(plainText: String) = plainText.trim.filter(_.isLetterOrDigit).toLowerCase

  def squareSize(plainText: String) = Math.ceil(Math.sqrt(plainText.length)).toInt

  def plaintextSegments(plainText: String) = {
    normalizePlaintext(plainText) match {
      case "" => Nil
      case normalized => normalized.grouped(squareSize(normalized)).toList
    }
  }

  def ciphertext(plainText: String) = transpose(plaintextSegments(plainText)).mkString

  def normalizedCiphertext(plainText: String) = transpose(plaintextSegments(plainText)).mkString(" ")
}

object CryptoSquare {
  def apply() = new CryptoSquare
}

File: javascript/crypto-square/crypto-square.js
export default class Crypto {
    constructor(input) {
        this.input = input;
    }

    size() {
        return Math.ceil(Math.sqrt(this.normalizePlaintext().length));
    }

    normalizePlaintext() {
        return this.input.toLowerCase().replace(/[^\w]/g, '');
    }

    normalizeCiphertext() {
        return this.transpose(this.plaintextSegments()).join(' ');
    }

    ciphertext() {
        return this.transpose(this.plaintextSegments()).join("");
    }

    plaintextSegments() {
        const segments = [];
        const plainText = this.normalizePlaintext();
        const segmentSize = this.size();
        const numberOfSegments = Math.ceil(plainText.length / segmentSize);

        for (let i = 0; i < numberOfSegments; i++)
            segments.push(plainText.substr(i * segmentSize, Math.min(plainText.length - i * segmentSize, segmentSize)));

        return segments;
    }

    transpose(input) {
        const transposed = [];

        for (let i = 0; i < input.length; i++) {
            for (let j = 0; j < input[i].length; j++) {
                if (transposed.length <= j)
                    transposed[j] = '';

                transposed[j] += input[i][j];
            }
        }

        return transposed
    }
}